File: src/exercise_topic_cpp/src/sensor_status_pub.cpp
#include "base_interfaces/msg/sensor_status.hpp"
#include "rclcpp/rclcpp.hpp"
#include <string>
#include <vector>

using namespace std::chrono_literals;

struct SensorData {
    std::string sensor_name;
    float temperature;
    float voltage;
    bool is_online;
};

class SensorStatusPub : public rclcpp::Node {
public:
    SensorStatusPub() : Node("sensor_status_publisher_cpp"), current_index_(0) {
        RCLCPP_INFO(this->get_logger(), "发布方创建成功!");

        sensor_data_ = {
            {"lidar", 45.0f, 12.0f, true},
            {"camera", 62.5f, 5.0f, true},
            {"imu", 38.0f, 3.3f, true},
            {"ultrasonic", 35.0f, 2.7f, true},
            {"gps", 40.0f, 0.0f, false},
        };

        publisher_ = this->create_publisher<base_interfaces::msg::SensorStatus>("sensor_status", 10);
        timer_ = this->create_wall_timer(1s, std::bind(&SensorStatusPub::timer_callback, this));
    }

private:
    void timer_callback() {
        const auto &sensor_data = sensor_data_[current_index_++];
        if (current_index_ >= sensor_data_.size()) {
            current_index_ = 0;
        }

        auto sensor_msg = base_interfaces::msg::SensorStatus();
        sensor_msg.sensor_name = sensor_data.sensor_name;
        sensor_msg.temperature = sensor_data.temperature;
        sensor_msg.voltage = sensor_data.voltage;
        sensor_msg.is_online = sensor_data.is_online;

        publisher_->publish(sensor_msg);
        RCLCPP_INFO(
            this->get_logger(),
            "发布传感器状态: [name=%s, temp=%0.1f, voltage=%0.1f, online=%s]",
            sensor_msg.sensor_name.c_str(),
            sensor_msg.temperature,
            sensor_msg.voltage,
            sensor_msg.is_online ? "true" : "false"
        );
    }

    std::vector<SensorData> sensor_data_;
    size_t current_index_;

    rclcpp::Publisher<base_interfaces::msg::SensorStatus>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<SensorStatusPub>());
    rclcpp::shutdown();

    return 0;
}

File: src/exercise_topic_py/exercise_topic_py/sensor_status_pub.py
import rclpy
from rclpy.node import Node
from base_interfaces.msg import SensorStatus

class SensorStatusPub(Node):
    def __init__(self):
        super().__init__("sensor_status_publisher_py")
        self.get_logger().info("发布方创建成功!(Python)")

        self.sensor_data_ = [
            ("lidar",       45.0, 12.0, True),
            ("camera",      62.5,  5.0, True),
            ("imu",         38.0,  3.3, True),
            ("ultrasonic",  35.0,  2.7, True),
            ("gps",         40.0,  0.0, False),
        ]
        self.current_index_ = 0

        self.publisher_ = self.create_publisher(SensorStatus, "sensor_status", 10)
        self.timer_ = self.create_timer(1.0, self.timer_callback)

    def timer_callback(self):
        sensor_name, temperature, voltage, is_online = self.sensor_data_[self.current_index_]

        status_msg = SensorStatus()
        status_msg.sensor_name = sensor_name
        status_msg.temperature = temperature
        status_msg.voltage = voltage
        status_msg.is_online = is_online

        self.publisher_.publish(status_msg)
        self.get_logger().info(
            f"发布传感器状态: [name={status_msg.sensor_name}, "
            f"temp={status_msg.temperature:.1f}, "
            f"voltage={status_msg.voltage:.1f}, "
            f"online={status_msg.is_online}]"
        )

        self.current_index_ = (self.current_index_ + 1) % len(self.sensor_data_)

def main():
    rclpy.init()
    rclpy.spin(SensorStatusPub())
    rclpy.shutdown()

if __name__ == "__main__":
    main()